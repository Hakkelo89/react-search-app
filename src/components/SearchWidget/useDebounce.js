import { useState, useEffect } from 'react'

const useDebounce = (searchTerm) => {
  const [value, setValue] = useState(searchTerm)

  useEffect(() => {
    const timer = setTimeout(() => {
      setValue(searchTerm)
    }, 500)

    return () => {
      clearTimeout(timer)
    }
  }, [searchTerm])
  
  return (value && value.length > 1) ? `https://www.rentalcars.com/FTSAutocomplete.do?solrIndex=fts_en&solrRows=10&solrTerm=${value}` : ''
}

export default useDebounce