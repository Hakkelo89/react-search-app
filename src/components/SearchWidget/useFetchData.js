import { useState, useEffect } from 'react'
import axios from 'axios'

const useFetchData = (searchURL) => {
  const [results, setResults] = useState([]) 
  
  useEffect(() => {
    if (searchURL) {
      const fetchData = async () => {
        const { data } = await axios.get(searchURL)
        const transformedData = data.results.docs.map(({ name }) => name)
        setResults(transformedData)
      }
      fetchData()
    } else setResults([])
  }, [searchURL])

  return results
}

export default useFetchData