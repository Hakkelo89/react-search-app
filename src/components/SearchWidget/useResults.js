import { useState } from 'react'

import useFetchData from './useFetchData'
import useDebounce from './useDebounce'

const useResults = (initialState = {}) => { 
  const {
    initialShowResult = false,
    initialSearchTerm = '',
  } = initialState
  const [showResults, setShowResults] = useState(initialShowResult)
  const [searchTerm, setSearchTerm] = useState(initialSearchTerm)

  const onBlur = () => setShowResults(false)

  const onFocus = () => setShowResults(true)

  const onChange = ({ target: { value } }) => setSearchTerm(value)

  const searchURL = useDebounce(searchTerm)
  
  const data = useFetchData(searchURL)

  const props = {
    onBlur,
    onFocus,
    onChange,
  }
  
  return [props, showResults, data]
}

export default useResults