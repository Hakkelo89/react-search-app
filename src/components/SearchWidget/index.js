import React from 'react'
import PropTypes from 'prop-types'

import useResults from './useResults'
import SearchResults from '../SearchResults'

const SearchWidget = ({ label, title, placeholder }) => {
  const [inputProps, showResults, results] = useResults()

  return (
    <div className="c-searchwidget">
      <h2 className="c-searchwidget__title">{title}</h2>
      <div className="c-searchwidget__searchbox">
        <label className="c-form-field__label">{label}</label>
        <div className="c-form-field__input">
          <input
            type="text"
            placeholder={placeholder}
            {...inputProps}
          />
          <SearchResults showResults={showResults} results={results} />
        </div>
      </div>
    </div>
  )
}

SearchWidget.defaultProps = {
  label: 'Pick-up Location',
  placeholder: 'city, airport, station, region, district...',
  title: 'Where are you going?',
}

SearchWidget.propTypes = {
  label: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
}

export default SearchWidget