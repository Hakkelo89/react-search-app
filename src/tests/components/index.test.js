import React from 'react'
import { shallow } from 'enzyme'
import SearchWidget from '../../components/SearchWidget'

const render = () => shallow(<SearchWidget />)

test('should render the snapshot of the SearchWidget component', () => {
  expect(render()).toMatchSnapshot();
})

test('should render the SearchWidget container', () => {
  expect(render().find('.c-searchwidget').exists()).toEqual(true)
})

test('should render the SearchWidget title', () => {
  expect(render().find('.c-searchwidget__title').exists()).toEqual(true)
})

test('should render the SearchWidget title with default title', () => {
  expect(render().find('.c-searchwidget__title').text()).toEqual('Where are you going?')
})

test('should render the searchbox container', () => {
  expect(render().find('.c-searchwidget__searchbox').exists()).toEqual(true)
})

test('should render the form field label', () => {
  expect(render().find('.c-form-field__label').exists()).toEqual(true)
})

test('should render the form field label with default label', () => {
  expect(render().find('.c-form-field__label').text()).toEqual('Pick-up Location')
})

test('should render the form-field input container', () => {
  expect(render().find('.c-form-field__input').exists()).toEqual(true)
})

test('should render an input element', () => {
  expect(render().find('input').exists()).toEqual(true)
})

test('should render an input element with default placeholder', () => {
  expect(render().find('input').prop('placeholder')).toEqual('city, airport, station, region, district...')
})

test('should render an input element with type \'text\'', () => {
  expect(render().find('input').prop('type')).toEqual('text')
})

test('should render the search results container', () => {
  expect(render().find('.c-search-results').exists()).toEqual(true)
})

test('should render the search results item with expected count of 3', () => {
  expect(render().find('.c-search-results__item').length).toEqual(3)
})
