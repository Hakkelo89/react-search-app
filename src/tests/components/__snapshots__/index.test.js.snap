// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the snapshot of the SearchWidget component 1`] = `
<div
  className="c-searchwidget"
>
  <h2
    className="c-searchwidget__title"
  >
    Where are you going?
  </h2>
  <div
    className="c-searchwidget__searchbox"
  >
    <label
      className="c-form-field__label"
    >
      Pick-up Location
    </label>
    <div
      className="c-form-field__input"
    >
      <input
        placeholder="city, airport, station, region, district..."
        type="text"
      />
      <div
        className="c-search-results"
      >
        <div
          className="c-search-results__item"
        >
          Manchester Airport
        </div>
        <div
          className="c-search-results__item"
        >
          Manchester
        </div>
        <div
          className="c-search-results__item"
        >
          Manchester - Piccadilly Train Station
        </div>
      </div>
    </div>
  </div>
</div>
`;
